// @antv/x6@3.1.7/lib/common/function/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throttle = exports.debounce = exports.call = exports.apply = exports.toDeferredBoolean = exports.toAsyncBoolean = exports.isAsyncLike = exports.isAsync = void 0;
var async_1 = require("./async");
Object.defineProperty(exports, "isAsync", { enumerable: true, get: function () { return async_1.isAsync; } });
Object.defineProperty(exports, "isAsyncLike", { enumerable: true, get: function () { return async_1.isAsyncLike; } });
Object.defineProperty(exports, "toAsyncBoolean", { enumerable: true, get: function () { return async_1.toAsyncBoolean; } });
Object.defineProperty(exports, "toDeferredBoolean", { enumerable: true, get: function () { return async_1.toDeferredBoolean; } });
var function_1 = require("./function");
Object.defineProperty(exports, "apply", { enumerable: true, get: function () { return function_1.apply; } });
Object.defineProperty(exports, "call", { enumerable: true, get: function () { return function_1.call; } });
Object.defineProperty(exports, "debounce", { enumerable: true, get: function () { return function_1.debounce; } });
Object.defineProperty(exports, "throttle", { enumerable: true, get: function () { return function_1.throttle; } });
